### DISCUSSION TITLE
'Plugin already registered' error when registering multiple plugins with `setuptools`

### DISCUSSION THREAD

**Discussion by dgard1981** [pytest-dev/pytest#11230] [Q&A]
I am developing some plugins locally that I hope to one day publish, so I'm registering them using `setuptools`. As I'm only testing at the moment, to install them I'm using the `--editable` option.

```python
pip install  --editable pytest_plugins/pytest_plugin1
pip install  --editable pytest_plugins/pytest_plugin2
```

If just either one of the plugins is installed, `pytest` works fine. In this case only `pytest_plugin2` is installed.

```
user@computer:~$ pytest tests -q
.                                                                                                                                                                                              [100%]
Output from Plugin 2.
1 passed in 0.00s
```

However, if both are installed `pytest` fails. It looks as though both plugins are trying to read from the same path, `'/home/user/pytest_plugins/pytest_plugin1/src/plugin.py`.

> ValueError: Plugin already registered under a different name: pytest_plugin2=<module 'src.plugin' from '/home/user/pytest_plugins/pytest_plugin1/src/plugin.py'>

I'm guessing this is somehow to do with how `pip` imports paths for `--editable` installs, but I'm not totally sure. Is this a known issue with plugins, and if so is there a workaround?

---

### Example files

```shell
.
├─ pytest_plugins
│   ├─ pytest_plugin1
│   │   ├─ src
│   │   │   └─ plugin.py
│   │   └─ setup.py
│   └─ pytest_plugin2
│       ├─ src
│       │   └─ plugin.py
│       └─ setup.py
└─ tests
    └─ test_something.py
```

<details>
  <summary><code>pytest_plugins/pytest_plugin1/setup.py</code></summary>
  
  ```python
from setuptools import setup, find_packages

setup(
    name='pytest_plugin1',
    packages=find_packages(include=['pytest_plugin1']),
    entry_points={'pytest11': ['pytest_plugin1 = src.plugin']}
)
  ```
</details>

<details>
  <summary><code>pytest_plugins/pytest_plugin1/src/plugin.py</code></summary>
  
  ```python
import pytest

@pytest.hookimpl
def pytest_terminal_summary(*_) -> None:
    print('Output from Plugin 1.')

  ```
</details>

<details>
  <summary><code>pytest_plugins/pytest_plugin2/setup.py</code></summary>
  
  ```python
from setuptools import setup, find_packages

setup(
    name='pytest_plugin2',
    packages=find_packages(include=['pytest_plugin2']),
    entry_points={'pytest11': ['pytest_plugin2 = src.plugin']}
)
  ```
</details>

<details>
  <summary><code>pytest_plugins/pytest_plugin2/src/plugin.py</code></summary>
  
  ```python
import pytest

@pytest.hookimpl
def pytest_terminal_summary(*_) -> None:
    print('Output from Plugin 2.')
  ```
</details>

<details>
  <summary><code>tests/test_something</code></summary>
  
  ```python
def test_something():
    assert 1 == 1
  ```
</details>

**Comment by dgard1981** [comment]
Further investigation shows that this is down to how the entry point is loaded. I'm note sure if the use of `--editable` packages has contributed to the issue.

```python
>>> import importlib_metadata
>>> for dist in list(importlib_metadata.distributions()):
...     for ep in dist.entry_points:
...         if ep.name.startswith('pytest_plugin'):
...             print(ep)
...             print(ep.load())
... 
EntryPoint(name='pytest_plugin1', value='src.plugin', group='pytest11')
<module 'src.plugin' from '/home/user/pytest_plugins/pytest_plugin1/src/plugin.py'>
EntryPoint(name='pytest_plugin2', value='src.plugin', group='pytest11')
<module 'src.plugin' from '/home/user/pytest_plugins/pytest_plugin1/src/plugin.py'>
```

**Comment by RonnyPfannschmidt** [comment]
All the plugins share the same module name ,thats bound to break, 


Different packages need different module top levels

**Reply by dgard1981** [reply]
But they don't share the same package name. The only similarity is the entry point within the package.

**Reply by RonnyPfannschmidt** [reply]
Both end up as src.plugin, package name doesn't matter, top-level import name does

**Comment by dgard1981** [comment]
I was able to resolve my issue with [here](https://github.com/pypa/setuptools/discussions/3989).
